src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -285,166 +299,224 @@ struct TreeVisitor<'tree> {
+            |error_kind| {
+                (self.err_builder)(ErrHandlerArgs {
+                    error_kind,
+                    conflicting_info: &this.nodes.get(idx).unwrap().debug_info,
+                    accessed_info: &this.nodes.get(self.initial).unwrap().debug_info,
+                })
+            },
+        )
+    }
+
+    fn go_upwards_from_accessed(
+        &mut self,
+        this: &mut TreeVisitor<'_>,
+        accessed_node: UniIndex,
+        visit_children: ChildrenVisitMode,
+    ) -> Result<(), OutErr> {
+        assert!(self.stack.is_empty());
+        // First, handle accessed node. A bunch of things need to
+        // be handled differently here compared to the further parents
+        // of `accesssed_node`.
+        {
+            self.propagate_at(this, accessed_node, AccessRelatedness::This)?;
+            if matches!(visit_children, ChildrenVisitMode::VisitChildrenOfAccessed) {
+                let accessed_node = this.nodes.get(accessed_node).unwrap();
+                for &child in accessed_node.children.iter().rev() {
+                    self.stack.push((child, AccessRelatedness::AncestorAccess, false));
                 }
             }
         }
+        // Then, handle the accessed node's parent. Here, we need to
+        // make sure we only mark the "cousin" subtrees for later visitation,
+        // not the subtree that contains the accessed node.
+        let mut last_node = accessed_node;
+        while let Some(current) = this.nodes.get(last_node).unwrap().parent {
+            self.propagate_at(this, current, AccessRelatedness::StrictChildAccess)?;
+            let node = this.nodes.get(current).unwrap();
+            for &child in node.children.iter().rev() {
+                if last_node == child {
+                    continue;
+                }
+                self.stack.push((child, AccessRelatedness::DistantAccess, false));
+            }
+            last_node = current;
+        }
+        self.stack.reverse();
         Ok(())
     }
 
-    fn new(initial: UniIndex, f_propagate: NodeApp, err_builder: ErrHandler) -> Self {
-        Self { initial, f_propagate, err_builder, stack: Vec::new() }
-    }
-
-    /// Finish the exploration by applying `exec_and_visit` until
-    /// the stack is empty.
-    fn finish(&mut self, visitor: &mut TreeVisitor<'_>) -> Result<(), OutErr> {
-        while let Some((idx, rel_pos)) = self.stack.pop() {
-            self.exec_and_visit(visitor, idx, /* no children to exclude */ None, rel_pos)?;
+    fn finish_foreign_accesses(&mut self, this: &mut TreeVisitor<'_>) -> Result<(), OutErr> {
+        while let Some((idx, rel_pos, is_final)) = self.stack.last_mut() {
+            let idx = *idx;
+            let rel_pos = *rel_pos;
+            if *is_final {
+                self.stack.pop();
+                self.propagate_at(this, idx, rel_pos)?;
+            } else {
```

Comment:
I think this is easier to follow if the order of the code matches the execution order. Also, please add comments.